agent/acs/handler/acs_handler.go

```diff
@@ -189,70 +189,61 @@ func NewSession(
 // Start starts the session. It'll forever keep trying to connect to ACS unless
 // the context is cancelled.
 //
-// If the context is cancelled, Start() would return with the error code returned
-// by the context.
-// If the instance is deregistered, Start() would emit an event to the
-// deregister-instance event stream and sets the connection backoff time to 1 hour.
+// If the context is cancelled returns nil otherwise returns error from context
 func (acsSession *session) Start() error {
-	// connectToACS channel is used to indicate the intent to connect to ACS
-	// It's processed by the select loop to connect to ACS
-	connectToACS := make(chan struct{}, 1)
-	// This is required to trigger the first connection to ACS. Subsequent
-	// connections are triggered by the handleACSError() method
-	connectToACS <- struct{}{}
+	// Loop continuously until context is closed/cancelled
 	for {
-		select {
-		case <-connectToACS:
-			seelog.Debugf("Received connect to ACS message")
-			// Start a session with ACS
-			acsError := acsSession.startSessionOnce()
-			select {
-			case <-acsSession.ctx.Done():
-				// agent is shutting down, exiting cleanly
+		seelog.Debugf("Attempting connect to ACS")
+		// Start a session with ACS
+		acsError := acsSession.startSessionOnce()
```

Comment:
From my previous comment, I think what what threw me off is that `startSessionOnce()` blocks internally, which is not very intuitive. I see how the code is equivalent now. 

Perhaps a more idiomatic way of doing this would be for `startSessionOnce()` to return a channel, instead of blocking internally. That way it's clear that the consumer (this function) can wait for the channel to return an error, or be closed. In that case, this method would look something like below, which in my opinion is more idiomatic/intutitive:

```golang
for {
    seelog.Debugf("Attempting connect to ACS")
    select {
    case acsErr := <- acsSession.startSession():
        // all the checks currently made
        . . .
    case <- acsSession.ctx.Done(): 
        return nil
    }
}
```

Reply:
I believe it is more Go-idiomatic to have blocking functions, so that the caller can decide its concurrency, instead of a function spinning up goroutine on its own and returning a channel. I think it is the caller that should decide if it needs a goroutine, and sometimes it does not like in this case. I find Kevin's implementation simple and correct.

Reply:
I don't think that helps create more clarity. startSessionOnce is synchronous and blocks as long as the connection is open that is very easy to reason about with synchronous code like in this PR. IF you really wanted to wait on something it should be the trigger to create a new connection which could be dispatched quickly then go back to this select. As it stands all the "checks currently made" are also synchronous and block for minutes to hours. If that is the contents of the select case then it doesn't follow for idiomatic message pump style coding in Go or any other language for that matter.

```go
for {
    select {
    case acsErr := <- acsSession.startSession:
        seelog.Debugf("Attempting connect to ACS")
        go acsSession.startSessionOnce()
    case <- acsSession.ctx.Done(): 
        return nil
    }
}

startSessionOnce() {
    // all code in startSessionOnce
   

   // after Serve returns but inside this fuction
   // all the checks currently made in select
   // if a time is selected
   go waitThenSignalChan(timeToWait, context, acsSession.startSession)
   // else if start immediately
   acsSession.startSession <- struct{}
}
```
This would be the idiomatic select usage but forces additional asynchronous goroutines and still allows for an easy control flow change that results in no ACS session ever again thus making the channel mechanism a burden not a help.

Reply:
besides, a function starting a goroutine and returning a channel is leak-prone: what happens if the caller exited on canceled context and **never received** from the channel? In order for that extra goroutine not to leak (i.e. to be garbage collected), some extra code would be required, which is even more complicated. Sending to a channel without a select is a smell.

Reply:
That said this code is still better than the previous one so I'm fine with merging it as is.

Reply:
> besides, a function starting a goroutine and returning a channel is leak-prone: what happens if the caller exits on canceled context and never receives from the channel?

that's fine. In go, the best practice is that functions that return channels are responsible for closing it at the appropriate time. Methods that consume channels returned from methods are supposed to keep reading from that channel until closed.